Add artifacts interface

coruscating opened the pull request:
### Summary

This PR adds the artifacts interface following the design in https://github.com/Qiskit/rfcs/blob/master/0007-experiment-dataframe.md.

### Details and comments

- Added the `ArtifactData` dataclass for representing artifacts.
- Added `ExperimentData.artifacts()`, `.add_artifacts()`, and `delete_artifact()` for working with artifacts, which is stored in a thread safe list. Currently the `ScatterTable` and `CurveFitResult` objects are stored as artifacts, and experiment serialization data will be added in the future.
- Artifacts are grouped by type and stored in a compressed format so that there aren't a huge number of individual files for composite experiments. As such, this PR depends on https://github.com/Qiskit-Extensions/qiskit-ibm-experiment/pull/93 to allow `.zip` formats for uploading to the cloud service. Inside each zipped file is a list of JSON artifact files with the filename equal to their unique artifact ID. For composite experiments with `flatten_results=True`, all `ScatterTable` artifacts are stored in `curve_data.zip` in individual jsons and so forth.
- Added a how-to for artifacts and updated documentation to demonstrate dataframe objects like AnalysisResults and the ScatterTable (`dataframe.css` is for styling these tables).
- Deprecated accessing analysis results via numerical indices to anticipate removing the curve fit result from analysis results altogether in the next release.
- Fixed bug where `figure_names` were being duplicated in a copied `ExperimentData` object.

Example experiment with artifacts ([link](https://quantum.ibm.com/experiments/eaad518d-232f-4cab-b137-e480ff7f1cbb)):
![image](https://github.com/Qiskit-Extensions/qiskit-experiments/assets/3870315/a2929782-dfef-4535-b246-1167666ebfc9)

nkanazawa1989 reviewed (commented):
Thanks @coruscating this looks good. Do you plan to add a tutorial for artifact? You also need to update [unittest util](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/48c9e67a10f8fe797526e2c02b7ab6dd12d22610/test/extended_equality.py#L332-L336) for ExperimentData equality check.

coruscating commented on qiskit_experiments/framework/artifact_data.py:
```diff
@@ -0,0 +1,110 @@
+from qiskit_experiments.framework.json import ExperimentEncoder, ExperimentDecoder
+
+
+class ArtifactData:
```
Good idea, done in 8eeba4f.

coruscating commented on releasenotes/notes/experiment-artifacts-c481f4e07226ce9e.yaml:
```diff
@@ -0,0 +1,15 @@
+    An artifact class has been introduced to store long-form data generated by experiments.
+    The :class:`.CurveFitResult` and :class:`.ScatterTable` generated by experiments 
+    are now stored in artifacts in the :class:`.ExperimentData` class. :meth:`.add_artifact`
+    and ?? have been added to manipulate the artifacts. These will be uploaded
```
Oops, forgot to write delete there, fixed 😅

coruscating commented on qiskit_experiments/framework/artifact_data.py:
```diff
@@ -0,0 +1,110 @@
+    for saving and retrieving to and from the experiments service, which stores this
+    as artifacts.
+
+    Types of objects that may be converted to artifacts include figures, fitted and raw data,
```
No, updated.

coruscating commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -2578,73 +2552,99 @@ def __str__(self):
-    def __json_decode__(cls, value):
-        return cls.__members__[value]  # pylint: disable=unsubscriptable-object
+        for key in artifact_keys:
+            del self._artifacts[key]
```
There's no upstream method to delete files—it was probably not needed before. I'll open a separate issue for this.

coruscating replied:
In fc058a9 I updated the behavior to overwrite an artifact file in the cloud with a size 0 file if all artifacts of that name have been deleted. This can be updated later to directly delete the artifact when that feature is added.

coruscating commented:
@nkanazawa1989 I've updated the PR, thanks. Here is a performance test for total time elapsed versus number of calls to the artifact upload API:
![image](https://github.com/Qiskit-Extensions/qiskit-experiments/assets/3870315/8b8b98e0-0d35-4e90-a176-c6e60299a833)
About half a second per call would make large experiments very slow, so I've switched to uploading a zipped file of each artifact type that contains json files corresponding to artifacts of that type, including ones from child experiments if `flatten_results` is on. By default, experiments have two artifacts, `curve_data.zip` and `fit_data.zip`. For this reason, `auto_save` doesn't update artifacts or it would lead to many API calls again.

coruscating commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -2578,73 +2552,99 @@ def __str__(self):
+        out = []
+        for key in artifact_keys:
+            artifact_data = self._artifacts[key]
+            if artifact_data is None:
```
No idea...I'll remove it.

coruscating commented on docs/conf.py:
```diff
@@ -236,6 +236,11 @@ def maybe_skip_member(app, what, name, obj, skip, options):
         "filter_kwargs",
         "fit_func",
         "signature",
+        "artifact_id",
+        "artifact_data",
+        "device_components",
+        "created_time",
+        "data",
```
This is the hacky way to fix sphinx build errors that was introduced in #983.

nkanazawa1989 commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -2268,6 +2223,18 @@ def load(
+                    if service.experiment_has_file(experiment_id, filename):
+                        artifact_file = service.file_download(experiment_id, filename)
+                        for artifact_string in zip_to_objs(artifact_file):
+                            artifact = json.loads(artifact_string, cls=cls._json_decoder)
```
Same here. Can this be `zip_to_objs(..., decoder: Type[JSONDecoder]) -> list[Any]`?

nkanazawa1989 commented on qiskit_experiments/database_service/utils.py:
```diff
@@ -74,6 +75,47 @@ def utc_to_local(utc_dt: datetime) -> datetime:
+                data = file.read().decode("utf-8")
+                output_strings.append(data)
+
+    return output_strings
```
I think it's better to yield here for memory footprint because each object can be heavy in artifact.

nkanazawa1989 reviewed (commented):
Thanks @coruscating , almost close. Just added further minor suggestions for machinery.

coruscating commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -1785,6 +1710,29 @@ def save(
+                        )
+                        self.service.file_upload(
+                            experiment_id=self.experiment_id,
+                            file_name=f"{file_type}.zip",
```
Right, that's better, and the current implementation wouldn't work properly when `delete_artifacts` is actually implemented. Done in 543ce53.

coruscating replied:
Moved metadata update to before saving the metadata in `save()`.

coruscating commented on qiskit_experiments/framework/containers/artifact_data.py:
```diff
@@ -0,0 +1,89 @@
+    def __json_encode__(self):
+        return {
+            "name": self.name,
+            "data": json.dumps(self.data, cls=self._json_encoder),
```
You're right, for some reason it wasn't working without this earlier but I must have had a bug.

nkanazawa1989 reviewed (commented):
Thanks, this looks good in code-wise, but found several malfunctions.

- Data cannot be retrieved

```python
from qiskit_experiments.library import T1
import numpy as np
from qiskit_experiments.framework import ExperimentData

t1exp = T1((0,), np.linspace(0, 300e-6, 11), backend=backend)
t1exp_data = t1exp.run().block_for_results()
t1exp_data.artifacts()  # two entries
t1exp_data.save()

ret_data = ExperimentData.load(t1exp_data.experiment_id, provider=provider)
ret_data.artifacts()  # empty
```

- CompositeCurveAnalysis doesn't return artifact

```python
from qiskit_experiments.library import CrossResonanceHamiltonian

exp = CrossResonanceHamiltonian((0,1), backend=backend)
exp_data = exp.run().block_for_results()
exp_data.artifacts()  # empty
```

I'll approve this PR after these two bugs are fixed and tests are added.

(edit) 

One more comment. Currently the file extension shown in the webpage is `*.zip.json` but I cannot open the file because it's broken. Probably the correct extension is `.zip`?

nkanazawa1989 commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -416,11 +421,11 @@ def _run_analysis(
             curve_data_set.append(table)
 
-        combined_curve_data = pd.concat(curve_data_set)
+        combined_curve_data = ScatterTable.from_dataframe(
```
You need to return this object and `fit_dataset` (or individual item in the dict) as ArtifactData.

coruscating replied:
Thanks for catching this. Fixed in 2052a24.

coruscating replied:
Oops, 2052a24 had an error but 120326f fixes it. `curve_data` contains the full multi-analysis `ScatterTable` while `fit_summary` is a dictionary of `CurveFitResult`s keyed by analysis, and loading works. Example `CrossResonanceHamiltonian` experiment: https://quantum.ibm.com/experiments/9f4b8f5f-dcad-40b2-88b3-351df2cdd379

coruscating commented on qiskit_experiments/database_service/utils.py:
```diff
@@ -74,6 +75,47 @@ def utc_to_local(utc_dt: datetime) -> datetime:
+
+    with zipfile.ZipFile(zip_buffer, "w", zipfile.ZIP_DEFLATED, False) as zip_file:
+        for file_name, data in zip(filenames, contents):
+            zip_file.writestr(file_name, data)
```
Thanks for suggesting this. I tried adding `zlib.compress(zip_buffer.read())` to the end of the method. For a 27 Q parallel T1 experiment, it only reduced `curve_data.zip` from 124 KB to 122 KB, and since this is more inconvenient for the user to decompress directly from the IQP interface, I'm not sure it's worth it. I also tried compressing each JSON file individually but the performance was even worse.

nkanazawa1989 commented:
Sorry I was using old ibm experiment package. I updated it and confirmed save/load work fine. In the previous version file extension became `zip.json` which doesn't match with the file names in the experiment metadata.

nkanazawa1989 commented on test/curve_analysis/test_baseclass.py:
```diff
@@ -226,12 +231,43 @@ def test_end_to_end_single_function(self):
         self.assertExperimentDone(result)
 
+        curve_data = result.artifacts("curve_data").data
+        np.testing.assert_array_equal(curve_data.series_name, "test")
```
I guess this returns array. Does assert_array_equal broadcast second value?

coruscating replied:
It broadcasts scalars and raises an assertion if the two objects are not equal: https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html
I just used it since we already use it in this manner in tests.

nkanazawa1989 commented on test/curve_analysis/test_baseclass.py:
```diff
@@ -226,12 +231,43 @@ def test_end_to_end_single_function(self):
+        curve_data = result.artifacts("curve_data").data
+        np.testing.assert_array_equal(curve_data.series_name, "test")
+        np.testing.assert_array_equal(curve_data.analysis, "CurveAnalysis")
+        np.testing.assert_array_equal(len(curve_data.filter(category="raw")), 100)
```
```suggestion
        self.assertEqual(len(curve_data.filter(category="raw")), 100)
```
Why testing with assert_array_equal?

coruscating replied:
Oops, copy paste error, thanks.

nkanazawa1989 commented on test/curve_analysis/test_baseclass.py:
```diff
@@ -226,12 +231,43 @@ def test_end_to_end_single_function(self):
         result = analysis.run(test_data)
         self.assertExperimentDone(result)
 
+        curve_data = result.artifacts("curve_data").data
```
Alternatively you can create reference dataframe and directly compare with
```python
ref_data = ScatterTable.from_dataframe(ref_frame)
self.assertEqual(curve_data, ref_data)
```
I guess this is more strict.

coruscating replied:
That's true, but the comparison can't be made directly because the `ScatterTable` contains random `yval` and `yerr` values that would have to be excluded from the comparison (unless we save the full `ScatterTable` for a particular seed in the test code), so I thought the two different approaches would be about the same.

nkanazawa1989 replied:
Here is the code. You can run data_processor locally. 

```python
        curve_data = result.artifacts("curve_data").data

        processed_y = data_processor(test_data.data())
        ref_raw_subset = pd.DataFrame.from_dict(
            {
                "xval": x,
                "yval": unp.nominal_values(processed_y),
                "yerr": unp.std_devs(processed_y),
                "series_name": np.full(100, "test"),
                "series_id": np.full(100, 0),
                "category": np.full(100, "raw"),
                "shots": np.full(100, 10000),
                "analysis": np.full(100, "CurveAnalysis"),
            }
        )
        self.assertEqual(
            ScatterTable.from_dataframe(ref_raw_subset),
            curve_data.filter(category="raw"),
        )
```

But I found that it's difficult to test `formatted` and `fitted` subset. Currently test is performed against everything we can predict, but this is likely just testing the behavior of ScatterTable itself. In the context of testing curve analysis, probably we need to test `CurveAnalysis._format_data` and fit curve calculation in `CurveAnalysis._run_analysis`, i.e. y values of formatted and fitted subset. Since these calculation is done behind the scene and preparation of the reference data is bit tough. For now we can go as-is and update the test when we come up with nice solution.

nkanazawa1989 commented on test/curve_analysis/test_baseclass.py:
```diff
@@ -629,12 +735,52 @@ def test_multi_composite_curve_analysis(self, plot, gen_figures, n_figures):
+        self.assertAlmostEqual(fit_B.params["freq"], freq2, delta=0.1)
+        self.assertAlmostEqual(fit_B.params["b"], b2, delta=0.1)
+
+        table_subset = expdata.artifacts("curve_data").data.filter(series=series, category=category)
```
How the setting for the figure generation impacts the ScatterTable?

coruscating replied:
It doesn't, I just tested both here since it's the integration test for composite curve analysis.

nkanazawa1989 commented on test/database_service/test_db_experiment_data.py:
```diff
@@ -1201,3 +1200,90 @@ def test_hgp_setter(self):
+        # delete by name
+        exp_data.delete_artifact("test")
+        self.assertEqual(exp_data.artifacts(), [])
+        self.assertEqual(exp_data._deleted_artifacts, {"test"})
```
Do we want to test against protected method? This means we need to maintain the behavior of this method.

coruscating replied:
Yeah it's not meant for external use, just like `exp_data._deleted_figures`, but I want to test that it has the correct behavior because there's no easy way to check that deletion worked correctly on the fake service (I think). We can update the test when the fake service is updated.

nkanazawa1989 reviewed (commented):
Thanks @coruscating . Just few minor comments. Overall this looks good to me.

coruscating commented on docs/tutorials/curve_analysis.rst:
```diff
@@ -240,6 +240,85 @@ generate initial guesses for parameters, from the ``AnalysisA`` class in the fir
+without an overhead of complex data management, as well as end-users with
+retrieving and reusing the intermediate data for their custom fitting workflow
+outside our curve fitting framework.
+Note that a :class:`ScatterTable` instance may be saved in the :class:`.ExperimentData` as an artifact.
```
Already on the next line ;)

coruscating commented on qiskit_experiments/database_service/utils.py:
```diff
@@ -74,6 +75,52 @@ def utc_to_local(utc_dt: datetime) -> datetime:
+    with zipfile.ZipFile(zip_buffer, "r") as zip_file:
+        for file_name in zip_file.namelist():
+            with zip_file.open(file_name) as file:
+                json_data = file.read().decode("utf-8")
```
Actually both `encode()` and `decode()` default to `utf-8`, so this is not necessary either way.

wshanks commented on releasenotes/notes/experiment-artifacts-c481f4e07226ce9e.yaml:
```diff
@@ -0,0 +1,20 @@
+  - |
+    Direct access to the curve fit summary in :class:`.ExperimentData` has moved from 
+    :meth:`.analysis_results` to :meth:`.artifacts`, where values are stored in
+    :meth:`.artifacts.data`. For example, to access the chi-squared of the fit,
```
Does this work? It looks funny that it renders as `data()` in this line and then `data.chisq` on the next line.

```suggestion
    :attr:`.artifacts.data`. For example, to access the chi-squared of the fit,
```

coruscating replied:
Thanks for catching this, updated.

wshanks commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -2364,6 +2378,10 @@ def copy(self, copy_results: bool = True) -> "ExperimentData":
 
+        with self._artifacts.lock:
+            new_instance._artifacts = ThreadSafeOrderedDict()
+            new_instance.add_artifacts(self._artifacts.values())
```
We don't do enough threading for it to matter but `ExperimentData` should probably have a top level lock. Ideally, all the data should be copied within a single lock instead of one per subcomponent.

wshanks commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -2602,75 +2621,101 @@ def __str__(self):
         return ret
 
+    def add_artifacts(self, artifacts: ArtifactData | list[ArtifactData], overwrite: bool = False):
+        """Add artifacts of experiment. The artifact ID must be unique.
```
```suggestion
        """Add artifacts to experiment. The artifact ID must be unique.
```

wshanks reviewed (commented):
This looks good. I have a couple of concerns to get your thoughts on. I also commented on several non-blocking things that could be turned into new issues.

coruscating commented on qiskit_experiments/curve_analysis/base_curve_analysis.py:
```diff
@@ -214,7 +207,7 @@ def _default_options(cls) -> Options:
         options.plotter = CurvePlotter(MplDrawer())
         options.plot_raw_data = False
-        options.return_fit_parameters = True
+        options.return_fit_parameters = False
```
Yeah, ideally I would like to deprecate accessing results by number entirely. I also mentioned up above that the order of artifacts will change when you save and load a composite experiment: https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1342#issuecomment-1920926420.

In this case, I can shift simple indices by 1, but `analysis_results()` accepts slices and we'd have to introduce a bunch of code checking if the slice includes 0 and how to shift it. I'm not sure what the best solution is—the simplest is just to give a warning when users try to use numerical indices.

wshanks replied:
I don't think you want to shift all indices because only CurveAnalysis indices changed. That's most things but some like tomography do not use CurveAnalysis and would not have shifted.

Maybe we should not accept `int` for `ExperimentData.artifacts()`?

wshanks replied:
By the way, based on #1385 it seems like we are treating the moving of the fit parameters from analysis results to artifacts as a minor non-core change. Does that seem right?

coruscating replied:
Oh I think I misunderstood you. I thought you wanted to put all the logic in `expdata.analysis_results()` but it would be easier to just add a placeholder in the actual `AnalysisResultTable`. Would the placeholder be actual data or a blank row? I think if we're going to get rid of numerical indices across the package, we could be more aggressive in making the breaking change now to get users used to the idea of not using them.

wshanks replied:
Yes, I was thinking something mostly blank. It could have a name like  `@Parameters_Placeholder` or something like that and `ExperimentData.analysis_results(0)` could error if the result being returned were `@Parameters_Placeholder`. I think that is the only way the user gets a result directly?

coruscating replied:
Then I think it's better to just put back the original behavior and warn the user if they try to use `analysis_results(0)` or numerical indices. I think it would confuse new users why there's a placeholder there with no useful information. One issue with moving to non-numerical indices is that we don't make the analysis result name very easy to find in the docs or regularized. Some are styled like `Local Readout Mitigator` and some are `d_theta`.

nkanazawa1989 commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -2602,75 +2621,101 @@ def __str__(self):
+    def artifacts(
+        self,
+        artifact_key: int | str = None,
+    ) -> ArtifactData | list[ArtifactData]:
```
To be fair this pattern appears everywhere in experiments. I agree with your point though. I would cleanup the entire interface in next release.

wshanks replied:
Fair. It's avoiding adding opportunity for deprecated behavior vs keeping consistency across the API.

coruscating commented on test/database_service/test_db_experiment_data.py:
```diff
@@ -1064,6 +1063,19 @@ def test_copy_metadata(self):
+        for i in exp_data._figures.keys():
+            self.assertEqual(exp_data._figures[i], copied._figures[i])
+        for i in exp_data._artifacts.keys():
+            self.assertEqual(exp_data._artifacts[i], copied._artifacts[i])
```
Right, that's better and I found a bug where `figure_names` were being duplicated in the copied object.

coruscating commented on qiskit_experiments/framework/containers/artifact_data.py:
```diff
@@ -0,0 +1,66 @@
+
+    def __eq__(self, value):
+        """Test equality between two instances of ArtifactData."""
+        return vars(self) == vars(value)
```
Good catch, I think this was back when the class was unnecessarily complicated and custom equality was needed. It works fine without it now.

coruscating commented on docs/howtos/artifacts.rst:
```diff
@@ -0,0 +1,140 @@
+    fig, (ax1, ax2) = plt.subplots(1, 2)
+    ax1.errorbar(raw_data.x, raw_data.y, yerr=raw_data.y_err, capsize=5)
+    ax1.set_title(f"Raw data, {exp_type} experiment on {component}")
+    ax2.errorbar(fitted_data.x, fitted_data.y, yerr=fitted_data.y_err, capsize=5)
```
Yeah, I guess the motivation was that the default figure already does that, so this was an attempt to visualize the data differently, but it's definitely pretty artificial. I'll put it back into one figure with different styling from the default.

coruscating commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -2284,6 +2287,17 @@ def load(
+            if "artifact_files" in expdata.metadata:
+                for filename in expdata.metadata["artifact_files"]:
+                    if service.experiment_has_file(experiment_id, filename):
+                        artifact_file = service.file_download(experiment_id, filename)
```
Yeah, that's follow-up work for the cloud service.

coruscating commented on docs/howtos/artifacts.rst:
```diff
@@ -0,0 +1,140 @@
+    print("experiment_notes.zip")
+    print(f"|- {data.artifacts('experiment_notes').artifact_id}.json")
+
+Note that for performance reasons, the auto save feature does not apply to artifacts. You must still
```
Yeah, I think it would be nice to change auto_save to do a single upload of figures/results/artifacts in bulk once all analysis callbacks are complete as opposed to individual calls. Artifacts is the worst to upload one by one because of the zipped files but the others have room for performance improvement too.

coruscating commented on docs/howtos/artifacts.rst:
```diff
@@ -0,0 +1,140 @@
+Viewing artifacts
+~~~~~~~~~~~~~~~~~
+
+Here we run a parallel experiment consisting of two :class:`.T1` experiments in parallel and then view the output
```
It's the same behavior as analysis results and figures (added to the parent list of objects if `flatten_results` is turned on). I added a clarification in fe7485f.

coruscating commented on qiskit_experiments/framework/containers/artifact_data.py:
```diff
@@ -0,0 +1,66 @@
+
+@dataclass
+class ArtifactData:
+    """A dataclass for experiment data payloads.
```
Updated in fe7485f.

wshanks commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -1599,6 +1545,21 @@ def analysis_results(
             )
         self._retrieve_analysis_results(refresh=refresh)
 
+        if index == 0:
```
I think you want to check that index 0 holds fit parameters and otherwise fall back to the other branch here.

coruscating replied:
Good catch. Added a check that the name starts with `@` since importing `PARAMS_ENTRY_PREFIX` would be circular and we already use starting with `@` as a filtering criteria when sending data to the plotter.

wshanks reviewed (approved):
Looks good to me! I will try to make issues from my other comments.
